_posts/2016-04-04-pattern-matching.md

```diff
@@ -0,0 +1,120 @@
+---
+layout: post
+title: "Bring Sanity to Your Data Structures with Elixr Pattern Matching"
+description: "Examples of pattern matching deeply nested data structures in Elixir."
+category:
+tags: [elixir, pattern matching, functional programming, api]
+author: vikram_ramakrishnan
+---
+{% include JB/setup %}
+
+This short (I promise) blog post offers an alternate method of extracting data
+from a deeply nested data structure using Elixir pattern matching.
+
+## Pattern Matching
+
+Elixir's docs on [pattern matching](http://elixir-lang.org/getting-started/pattern-matching.html#pattern-matching) are brief, which is a shame because it is a powerful and intuitive
+way to extract information you need if you know the *precise* format of an
+incoming data structure. Most examples of pattern matching contain relatively
+innocuous examples like these:
+
+```elixir
+[x, y] = [1, 10]
+```
+
+or
+
+```elixir
+{x, y} = {1, 2}
+```
+
+Both of these match `x` with `1` and `y` with `10`. Seems straightforward, right?
+What about more complicated structures? Can we pattern match on nested data
+structures? Let's take a look:
+
+```elixir
+[[], _, [_, _, canyoufindme]] = [[], "anything", ["oh", "hai", "thereyouare"]]
+```
+
+Here, we match `canyoufindme` with the string `"thereyouare"`. By the way,
+the `_` simply ignores whatever's on the right hand side of the match. `_`
+just means 'I don't care about assigning the right side of this match
+to anything so ignore me.' 
+
+Without pattern matching, we'd end up doing something ridiculous (and
```

Comment:
Maybe we __might__ do something ridiculous like this vs. we would. I'm not sure this would be a way anybody would try. There are function in `List` for doing this much more simply:

```elixir
iex(1)> list = [[], "anything", ["oh", "hai", "thereyouare"]]
[[], "anything", ["oh", "hai", "thereyouare"]]
iex(2)> list |> List.last |> List.last
"thereyouare"
iex(3)>
```